packages/next-swc/crates/core/src/relay.rs

```diff
@@ -0,0 +1,284 @@
+use pathdiff::diff_paths;
+use regex::{Captures, Regex};
+use relay_compiler_common::SourceLocationKey;
+use std::borrow::Cow;
+use std::path::PathBuf;
+
+use relay_compiler::compiler_state::{SourceSet, SourceSetName};
+use relay_compiler::{create_path_for_artifact, FileCategorizer, FileGroup, ProjectConfig};
+use serde::Deserialize;
+use swc_atoms::JsWord;
+use swc_common::errors::HANDLER;
+use swc_common::FileName;
+use swc_ecmascript::ast::*;
+use swc_ecmascript::utils::{quote_ident, ExprFactory};
+use swc_ecmascript::visit::{Fold, FoldWith};
+
+#[derive(Copy, Clone, Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum RelayLanguageConfig {
+    Typescript,
+    Flow,
+}
+
+struct Relay {
+    file_name: FileName,
+    relay_config_for_tests: Option<ProjectConfig>,
+}
+
+fn pull_first_operation_name_from_tpl(tpl: &TaggedTpl) -> Option<String> {
+    tpl.tpl
+        .quasis
+        .iter()
+        .filter_map(|quasis| {
+            let regex = Regex::new(r"(fragment|mutation|query|subscription) (\w+)").unwrap();
+            let capture_group = regex.captures_iter(&quasis.raw.value).next();
+
+            match capture_group {
+                None => None,
+                Some(capture_group) => {
+                    let operation_name: &str = &capture_group[2];
+
+                    Some(operation_name.to_string())
+                }
+            }
+        })
+        .next()
+}
+
+fn build_require_expr_from_path(path: String) -> Expr {
```

Comment:
not directly related to this specific PR, but I wonder how this can be implemented here. 

Currently, our `babel` plugin fundamentally does two things: 1) adds require('...') calls instead of `graphql` tags. and 2) warns if the text operation/fragment is changed, but was not recompile, and displays a runtime warning: https://github.com/facebook/relay/blob/main/packages/babel-plugin-relay/compileGraphQLTag.js#L128

Not sure if there is JS runtime part in next.js that can also handle that second part.

Reply:
@alunyov Can you explain the use case for 2? Might be able to get away with less complexity if we don't add it.

Reply:
@tbezman Sure! A developer changed the fragment or a query (added/changed field, etc), but didn't run the `relay-compiler`. When they refresh the page, the generated artifact for this file (.graphql.js file) won't contain the changes, this sometimes may lead to confusing situations, when developers may not see changes they are expecting on the screen.

To reduce the confusion, In the babel plugin we're comping the `md5` hash of the text inside `graphql` tag. And we're comparing this hash to the `hash` in the generated artifact (created by the compiler), if these hashes are different - we're showing a warning in the browser's console. When developers debug the outdated views, they may see the reason in the console.

Reply:
That makes a ton of sense. @alunyov Is there somewhere in the relay source where y'all export a function to hash a query / mutation / fragment / subscription? Hoping we can avoid duplicating functionality here.

Reply:
returns a value referencing data owned by the current function

Reply:
Also, maybe we can address this in a follow up PR. I'm helping a team migrate to Relay / GraphQL this week and want to keep them on the Rust compiler.

Reply:
@tbezman sure, totally fine to do this in a follow-up.

This is what we're using to create these `source_hashes`: 

https://github.com/facebook/relay/blob/14550c8bd80e53369852db66d7681f13d6dd5f84/compiler/crates/relay-compiler/src/build_project/build_ir.rs#L33-L36